php/libraries/ParentIDGenerator.php

```diff
@@ -0,0 +1,55 @@
+<?php declare(strict_types=1);
+/**
+ * This file contains a class used to generate ParentIDs for use in LORIS.
+ *
+ * PHP Version 7
+ *
+ * @category Main
+ * @package  LORIS
+ * @author   Shen Wang <wangshen.mcin@gmail.com>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+namespace LORIS\StudyEntities\Candidate\ParentIDGenerator;
+/**
+ * This class defines functions used to generate valid ParentID identifiers in
+ * LORIS.
+ *
+ * PHP Version 7
+ *
+ * @category Main
+ * @package  LORIS
+ * @author   Shen Wang <wangshen.mcin@gmail.com>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class ParentIDGenerator extends \CandIDGenerator
+{
+
+    protected $prefix ;
+
+    /**
+     * Creates a new ParentIDs generator by initializing properties based on class
+     * constants defined above.
+     *
+     * @param string $prefix the prefix string
+     */
+    public function __construct(string $prefix="Parent")
+    {
+        parent::__construct();
+        $this->prefix = $prefix;
+    }
+
+    /**
+     * Creates a new ParentID.
+     *
+     * @return string The new identifier.
+     */
+    public function generateParentID(): string
+    {
+        //  $generator = new \CandIDGenerator();
+        $candID = parent::generate();
+        return $this->prefix.(string)$candID;
```

Comment:
i dont like this ... the ParentID shouldn't have anything to do with the CandID. are you essentially using the CandIDGenerator to utilize the fact that it creates a 6 digit number?

Reply:
i think we should leave the pattern/sequence of the ParentID to be configurable by the project by supplying it in config.xml:

```
        <ParentID>
            <generation>random</generation>
            <structure>
                <seq type="static">PARENT</seq>
                <seq type="numeric" length="6" min='100000' max='999999'/>
            </structure>
        </ParentID>
```

Reply:
doing this will allow you to have a very simple ParentIDGenerator that looks like this:

```
class ParentIDGenerator extends SiteIDGenerator
{
    /**
     * Generates ParentIDs.
     *
     * @param ?string $prefix The site prefix to prepend to the ID value.
     *
     * @return void
     */
    public function __construct(?string $prefix = null)
    {
        $this->kind = 'ParentID';
        parent::__construct($prefix);
    }
}
```

calling ParentIDGenerator->generate() would just call the SiteIDGenerator generate() method which utilizes the IdentifierGenerator abstract class's createNewID() method